# Intro

> Generate [object_per_actor_nmvw.csv](object_per_actor_nmvw.csv)

```
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

SELECT (COUNT(DISTINCT ?nmvw_object) AS ?no_object) ?nmvw_actor
WHERE{
  GRAPH <https://pressingmatter.nl/NMVW/ccrdfobj.ttl>{
  {?nmvw_acq crm:P24_transferred_title_of ?nmvw_object .
   ?nmvw_acq crm:P23_transferred_title_from ?nmvw_actor .}
  UNION
  {
    ?nmvw_prod crm:P14_carried_out_by ?nmvw_actor .
    ?nmvw_object crm:P108i_was_produced_by ?nmvw_prod .
  }
  ?nmvw_object a crm:E22_Human-Made_Object .
  }
}GROUP BY ?nmvw_actor
```

In [92]:
def cleanup(x: str):
    return int(str(x).replace('"', '').strip())

import pandas as pd

df1 = pd.read_csv("object_per_actor_nmvw.csv")
df1 = df1.iloc[: , :-1]
df1.columns = ['no_object', 'nmvw_actor']
df1['no_object'] = df1['no_object'].apply(cleanup)
df1 = df1.sort_values(by=['no_object'], ascending=False)
df1

no_object                                       nmvw_actor
4564      135424    "https://hdl.handle.net/20.500.11840/pi4332" 
4048       69382   "https://hdl.handle.net/20.500.11840/pi40292" 
2608       13609    "https://hdl.handle.net/20.500.11840/pi3358" 
1448        9729    "https://hdl.handle.net/20.500.11840/pi2246" 
7559        7464   "https://hdl.handle.net/20.500.11840/pi57572" 
...          ...                                              ...
2558           1   "https://hdl.handle.net/20.500.11840/pi32984" 
2560           1    "https://hdl.handle.net/20.500.11840/pi3300" 
2561           1    "https://hdl.handle.net/20.500.11840/pi3305" 
7653           1   "https://hdl.handle.net/20.500.11840/pi57704" 
11954          1     "https://hdl.handle.net/20.500.11840/pi999" 

[11955 rows x 2 columns]

> Generate [object_per_actor_bb.csv](object_per_actor_bb.csv)
```
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>

SELECT (COUNT(DISTINCT ?bb_obj) AS ?no_object) ?bb_actor
WHERE{
    {?bb_acq crm:P23_transferred_title_from ?bb_actor .	
    ?bb_acq crm:P24_transferred_title_of ?bb_obj .}
    UNION
    {
          ?bb_obj crm:P51_has_former_or_current_owner ?bb_actor .
    }

GRAPH <https://pressingmatter.nl/Bronbeek/Objects/Objects/assertion/bad21d24/2024-03-01T12:54>{
    ?bb_obj a crm:E22_Human-Made_Object. }
}GROUP BY ?bb_actor
```
> Note: https://pressingmatter.nl/Bronbeek/Constituents/1896" --> "Koninklijk Tehuis voor Oud-Militairen en Museum Bronbeek"^^xsd:string
> https://pressingmatter.nl/Bronbeek/Constituents/9806 --> Nijmeegs Volkenkundig Museum
> https://pressingmatter.nl/Bronbeek/Constituents/9804 --> "Gemeente Nijmegen"
> https://pressingmatter.nl/Bronbeek/Constituents/9805 --> Jean Louis Henri Beijens

In [103]:
def cleanup(x: str):
    return int(str(x).replace('"', '').strip())

import pandas as pd

df1 = pd.read_csv("object_per_actor_bb.csv")
df1 = df1.iloc[: , :-1]
df1.columns = ['no_object', 'bb_actor']
df1['no_object'] = df1['no_object'].apply(cleanup)
df1 = df1.sort_values(by=['no_object'], ascending=False)
df1

no_object                                           bb_actor
2663      12409   "https://pressingmatter.nl/Bronbeek/Constitue...
6998       3066   "https://pressingmatter.nl/Bronbeek/Constitue...
6996       3066   "https://pressingmatter.nl/Bronbeek/Constitue...
6997       3066   "https://pressingmatter.nl/Bronbeek/Constitue...
4227       1542   "https://pressingmatter.nl/Bronbeek/Constitue...
322        1493   "https://pressingmatter.nl/Bronbeek/Constitue...
2345       1223   "https://pressingmatter.nl/Bronbeek/Constitue...
3348       1138   "https://pressingmatter.nl/Bronbeek/Constitue...
2950        800   "https://pressingmatter.nl/Bronbeek/Constitue...
2360        677   "https://pressingmatter.nl/Bronbeek/Constitue...

# CQ-1a

How many actors we have with one more objects and also have military background?

SPARQL Query:
```
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

SELECT (COUNT(DISTINCT ?nmvw_actor) AS ?no_actor)
WHERE{
  GRAPH <https://pressingmatter.nl/NMVW/ccrdfobj.ttl>{
  {?nmvw_acq crm:P24_transferred_title_of ?nmvw_object .
   ?nmvw_acq crm:P23_transferred_title_from ?nmvw_actor .}
  UNION
  {
    ?nmvw_prod crm:P14_carried_out_by ?nmvw_actor .
    ?nmvw_object crm:P108i_was_produced_by ?nmvw_prod .
  }
  ?nmvw_object a crm:E22_Human-Made_Object .
  }
  ?nmvw_actor owl:sameAs ?bb_actor .
}
```
> 277 actor from NMVW dataset, has connection with at least one object's provenance from NMVW dataset and also have owl:sameAS link with Bronbeek actor.
> Note that, here we consider `production actor` too, without this the value would be '204'.

# CQ 1b

In [ ]:
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

SELECT ?nmvw_object ?nmvw_actor
WHERE{
  GRAPH <https://pressingmatter.nl/NMVW/ccrdfobj.ttl>{
  {?nmvw_acq crm:P24_transferred_title_of ?nmvw_object .
   ?nmvw_acq crm:P23_transferred_title_from ?nmvw_actor .}
  UNION
  {
    ?nmvw_prod crm:P14_carried_out_by ?nmvw_actor .
    ?nmvw_object crm:P108i_was_produced_by ?nmvw_prod .
  }
  ?nmvw_object a crm:E22_Human-Made_Object .
  }
  # removing "onbeked / unknown"
  FILTER (?nmvw_actor != <https://hdl.handle.net/20.500.11840/pi4332>)
}

In [117]:
def cleanup(x: str):
    return int(str(x).replace('"', '').strip())

import pandas as pd

df1 = pd.read_csv("CQ1b_nmvw_obj.csv")
df1 = df1.iloc[: , :-1]
df1.columns = ['nmvw_object', 'nmvw_actor']
df1

nmvw_object  \
0        "https://hdl.handle.net/20.500.11840/128796"    
1         "https://hdl.handle.net/20.500.11840/22185"    
2         "https://hdl.handle.net/20.500.11840/22199"    
3         "https://hdl.handle.net/20.500.11840/22199"    
4         "https://hdl.handle.net/20.500.11840/22201"    
...                                                ...   
621439   "https://hdl.handle.net/20.500.11840/277225"    
621440   "https://hdl.handle.net/20.500.11840/277226"    
621441   "https://hdl.handle.net/20.500.11840/277227"    
621442   "https://hdl.handle.net/20.500.11840/277228"    
621443   "https://hdl.handle.net/20.500.11840/277229"    

                                             nmvw_actor  
0         "https://hdl.handle.net/20.500.11840/pi1125"   
1        "https://hdl.handle.net/20.500.11840/pi39586"   
2         "https://hdl.handle.net/20.500.11840/pi7008"   
3        "https://hdl.handle.net/20.500.11840/pi39586"   
4         "https://hdl.handle.net/20.500.11840/pi2354"   
...                                                 ...  
621439   "https://hdl.handle.net/20.500.11840/pi44822"   
621440    "https://hdl.handle.net/20.500.11840/pi2687"   
621441    "https://hdl.handle.net/20.500.11840/pi2246"   
621442    "https://hdl.handle.net/20.500.11840/pi2246"   
621443    "https://hdl.handle.net/20.500.11840/pi2246"   

[621444 rows x 2 columns]

In [125]:
# Group by 'nmvw_actor' and count unique 'nmvw_object' per group
unique_counts = df1.groupby('nmvw_actor')['nmvw_object'].nunique()

# Filter rows where 'nmvw_actor' has less than 2 unique 'nmvw_object'
filtered_df = df1[df1['nmvw_actor'].isin(unique_counts[unique_counts > 1].index)]

print(filtered_df['nmvw_object'].nunique())

444517


> out of 446,310 unique objects who have connection with actor, 444,517 can be linked with at least one other objects

Generate [CQ1b_bb_obj.csv](CQ1b_bb_obj.csv)

```
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

SELECT ?nmvw_actor (COUNT(DISTINCT ?bb_object) AS ?n_bb_object) 
WHERE{
    ?nmvw_actor owl:sameAs ?bb_actor .
    #acquisition event
    {
    	?bb_acq crm:P23_transferred_title_from ?bb_actor .   
    	?bb_acq crm:P24_transferred_title_of ?bb_object .
    }
  UNION
    {
      	?bb_object crm:P51_has_former_or_current_owner ?bb_actor .
    }
  GRAPH <https://pressingmatter.nl/Bronbeek/Objects/Objects/assertion/bad21d24/2024-03-01T12:54>{
    ?bb_object a crm:E22_Human-Made_Object .
    }
} GROUP BY ?nmvw_actor
```

In [134]:
import pandas as pd

df2 = pd.read_csv("CQ1b_bb_obj.csv")
df2 = df2.iloc[: , :-1]
df2.columns = ["nmvw_actor", "bb_object" ]
df2 = df2[df2['nmvw_actor'].isin(df1['nmvw_actor'])]
df2["bb_object"].nunique()

1766

> does any of the actor who have only one object, now have more than one connected object due to the connection of bronbeek,

In [137]:
df2[df1['nmvw_actor'].isin(unique_counts[unique_counts == 1].index)]['nmvw_actor'].nunique()

/var/folders/jg/8kz6nb554_3dz1_0swrmdw480000gn/T/ipykernel_1247/3693014615.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df2[df1['nmvw_actor'].isin(unique_counts[unique_counts == 1].index)]['nmvw_actor'].nunique()


7

# CQ 1c

How many possible object to object connection we have, by establishing links via actor?

Generate [CQ1b_nmvw_obj.csv](CQ1b_nmvw_obj.csv)

```
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

SELECT (COUNT(DISTINCT ?nmvw_object) AS ?n_nmvw_object) ?nmvw_actor
WHERE{
  GRAPH <https://pressingmatter.nl/NMVW/ccrdfobj.ttl>{
  {?nmvw_acq crm:P24_transferred_title_of ?nmvw_object .
   ?nmvw_acq crm:P23_transferred_title_from ?nmvw_actor .}
  UNION
  {
    ?nmvw_prod crm:P14_carried_out_by ?nmvw_actor .
    ?nmvw_object crm:P108i_was_produced_by ?nmvw_prod .
  }
  ?nmvw_object a crm:E22_Human-Made_Object .
  }
  # removing "onbeked / unknown"
  FILTER (?nmvw_actor != <https://hdl.handle.net/20.500.11840/pi4332>)
}GROUP BY ?nmvw_actor
```

> Note: https://hdl.handle.net/20.500.11840/pi40292 / "Stichting Indisch Wetenschappelijk Instituut" generate 69382 object connection but keeping this for now. 

In [110]:
def cleanup(x: str):
    return int(str(x).replace('"', '').strip())

import math

def calculate_combinations(total_objects):
    combination_value = math.factorial(total_objects) // (math.factorial(2) * math.factorial(total_objects - 2))
    return combination_value

import pandas as pd

df1 = pd.read_csv("CQ1b_nmvw_obj.csv")
df1 = df1.iloc[: , :-1]
df1.columns = ['n_nmvw_object', 'nmvw_actor']
df1['n_nmvw_object'] = df1['n_nmvw_object'].apply(cleanup)
df1 = df1[df1.n_nmvw_object != 1]
df1 = df1.sort_values(by=['n_nmvw_object'], ascending=False)
df1['possible_connections'] = df1['n_nmvw_object'].apply(calculate_combinations)
df1['possible_connections'].sum()


2918004358

In [111]:
df1

n_nmvw_object                                       nmvw_actor  \
4048          69382   "https://hdl.handle.net/20.500.11840/pi40292"    
2608          13609    "https://hdl.handle.net/20.500.11840/pi3358"    
1448           9729    "https://hdl.handle.net/20.500.11840/pi2246"    
7558           7464   "https://hdl.handle.net/20.500.11840/pi57572"    
1454           6603     "https://hdl.handle.net/20.500.11840/pi225"    
...             ...                                              ...   
9625              2   "https://hdl.handle.net/20.500.11840/pi60818"    
6244              2    "https://hdl.handle.net/20.500.11840/pi4888"    
9622              2   "https://hdl.handle.net/20.500.11840/pi60813"    
3873              2   "https://hdl.handle.net/20.500.11840/pi39864"    
7421              2   "https://hdl.handle.net/20.500.11840/pi57364"    

      possible_connections  
4048            2406896271  
2608              92595636  
1448              47321856  
7558              27851916  
1454              21796503  
...                    ...  
9625                     1  
6244                     1  
9622                     1  
3873                     1  
7421                     1  

[7467 rows x 3 columns]

How many object to object connection we have, by establishing links via actor when we also add objects from bronbeek via the owl:sameAs connection? 

Generate [CQ1b_bb_obj.csv](CQ1b_bb_obj.csv)
SPARQL QUERY:
```
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

SELECT ?nmvw_actor (COUNT(DISTINCT ?bb_object) AS ?n_bb_object) 
WHERE{
    ?nmvw_actor owl:sameAs ?bb_actor .
    #acquisition event
    {
    	?bb_acq crm:P23_transferred_title_from ?bb_actor .   
    	?bb_acq crm:P24_transferred_title_of ?bb_object .
    }
  UNION
    {
      	?bb_object crm:P51_has_former_or_current_owner ?bb_actor .
    }
  GRAPH <https://pressingmatter.nl/Bronbeek/Objects/Objects/assertion/bad21d24/2024-03-01T12:54>{
    ?bb_object a crm:E22_Human-Made_Object .
    }
} GROUP BY ?nmvw_actor

```

In [109]:
import pandas as pd

def cleanup(x: str):
    return int(str(x).replace('"', '').strip())

import math

def calculate_combinations(total_objects):
    combination_value = math.factorial(total_objects) // (math.factorial(2) * math.factorial(total_objects - 2))
    return combination_value

import pandas as pd

df2 = pd.read_csv("CQ1b_bb_obj.csv")
df2 = df2.iloc[: , :-1]
df2.columns = ["nmvw_actor" , "n_bb_object"]
df2['n_bb_object'] = df2['n_bb_object'].apply(cleanup)
df2 = df2.sort_values(by=['n_bb_object'], ascending=False)
df2


nmvw_actor  n_bb_object
277   "https://hdl.handle.net/20.500.11840/pi77221"          3066
146   "https://hdl.handle.net/20.500.11840/pi42066"           526
291   "https://hdl.handle.net/20.500.11840/pi80176"           367
292   "https://hdl.handle.net/20.500.11840/pi80287"           363
263    "https://hdl.handle.net/20.500.11840/pi7087"           243
..                                               ...          ...
162   "https://hdl.handle.net/20.500.11840/pi43035"             1
175   "https://hdl.handle.net/20.500.11840/pi43821"             1
176   "https://hdl.handle.net/20.500.11840/pi43826"             1
181   "https://hdl.handle.net/20.500.11840/pi44170"             1
325     "https://hdl.handle.net/20.500.11840/pi942"             1

[326 rows x 2 columns]

In [116]:
# Merge the dataframes on 'nmvw_actor'
merged_df = pd.merge(df1, df2, on='nmvw_actor', how='left')
merged_df['n_bb_object'] = merged_df['n_bb_object'].fillna(0)

# Add the 'n_nmvw_object' and 'n_bb_object' columns where 'nmvw_actor' matches
merged_df['combined_values'] = merged_df['n_nmvw_object'] + merged_df['n_bb_object']

merged_df['possible_connections'] = merged_df['combined_values'].apply(calculate_combinations)
merged_df['possible_connections'].sum()

/var/folders/jg/8kz6nb554_3dz1_0swrmdw480000gn/T/ipykernel_1247/1902449950.py:7: DeprecationWarning: Using factorial() with floats is deprecated
  combination_value = math.factorial(total_objects) // (math.factorial(2) * math.factorial(total_objects - 2))


2918355472

# CQ 2

SPARQL Query:
```
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

SELECT (COUNT(DISTINCT ?nmvw_object) AS ?n)
WHERE{
  Graph <https://pressingmatter.nl/NMVW/ccrdfobj.ttl>{
    {
     ?nmvw_acq crm:P23_transferred_title_from ?nmvw_actor .
     ?nmvw_acq crm:P24_transferred_title_of ?nmvw_object .
    }
    UNION
    {
      ?nmvw_object crm:P108i_was_produced_by ?nmvw_prod .
      ?nmvw_prod crm:P14_carried_out_by ?nmvw_actor .
    }
    ?nmvw_object  a crm:E22_Human-Made_Object .
  }
  ?nmvw_actor owl:sameAs ?bronbeek_actor .
}
```
> 18,551

# CQ 3

generate [CQ3_nmvw.csv](CQ3_nmvw.csv)

```
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

SELECT (COUNT(?nmvw_actor) AS ?no_actor) ?nmvw_object
WHERE{
  Graph <https://pressingmatter.nl/NMVW/ccrdfobj.ttl>{
     ?nmvw_acq crm:P23_transferred_title_from ?nmvw_actor .
     ?nmvw_acq crm:P24_transferred_title_of ?nmvw_object .
   }
}GROUP BY ?nmvw_object
```

In [143]:
def cleanup(x: str):
    return int(str(x).replace('"', '').strip())

import math

def calculate_combinations(total_objects):
    combination_value = math.factorial(total_objects) // (math.factorial(2) * math.factorial(total_objects - 2))
    return combination_value

import pandas as pd

df1 = pd.read_csv("CQ3_nmvw.csv")
df1 = df1.iloc[: , :-1]
df1.columns = [ "no_actor" , "nmvw_object"]
df1['no_actor'] = df1['no_actor'].apply(cleanup)
df1 = df1[df1.no_actor != 1]
df1 = df1.sort_values(by=['no_actor'], ascending=False)
df1['possible_connections'] = df1['no_actor'].apply(calculate_combinations)
df1['possible_connections'].sum()

162432

generate [CQ3_bb_before_links.csv](CQ3_bb_before_links.csv)

```
SELECT * WHERE{
     # ?nmvw_actor owl:sameAs ?bb_actor1
     ?bb_acq1 crm:P23_transferred_title_from ?bb_actor1 .
     ?bb_acq1 crm:P24_transferred_title_of ?bb_object .
    
    ?bb_acq2 crm:P24_transferred_title_of ?bb_object .
    ?bb_acq2 crm:P23_transferred_title_from ?bb_actor2 .

    Filter (?bb_actor1!= ?bb_actor2)
Graph <https://pressingmatter.nl/Bronbeek/Objects/Objects/assertion/bad21d24/2024-03-01T12:54>{
        ?bb_object a crm:E22_Human-Made_Object.
   }
}
```

In [156]:
import pandas as pd

df2 = pd.read_csv("CQ3_bb_before_links.csv")
df2 = df2.iloc[: , :-1]
df2.columns = [  "bb_acq1" , "bb_actor1" , "bb_object" , "bb_acq2" , "bb_actor2"]
# Count number of rows group by 'bb_object'
actor_count = df2.groupby('bb_object').size().reset_index(name='row_count')
actor_count['row_count'] = actor_count['row_count'].apply(lambda x: int(x/2))
print(actor_count['row_count'].sum())

7872


#### After Links

TODO: I can't get read of the duplicate events. 

Generate [CQ3_nmvw_obj.csv](CQ3_nmvw_obj.csv)
```
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

SELECT *
WHERE{
  Graph <https://pressingmatter.nl/NMVW/ccrdfobj.ttl>{
     ?nmvw_acq1 crm:P23_transferred_title_from ?nmvw_actor1 .
     ?nmvw_acq1 crm:P24_transferred_title_of ?nmvw_object .
    
    ?nmvw_acq2 crm:P24_transferred_title_of ?nmvw_object .
    ?nmvw_acq2 crm:P23_transferred_title_from ?nmvw_actor2 .
    
    Filter (?nmvw_actor1!= ?nmvw_actor2)
   }
}
```

In [179]:
import pandas as pd

df1 = pd.read_csv("CQ3_nmvw_obj.csv")
df1 = df1.iloc[: , :-1]
df1.columns = ["nmvw_acq1" , "nmvw_actor1" , "nmvw_object" , "nmvw_acq2" , "nmvw_actor2"]

pairs = set(frozenset(pair) for pair in zip(df1['nmvw_actor1'], df1['nmvw_actor2']))

# Filter rows from df1 where reversed pairs are not present
# df1_filtered = df1[df1.apply(lambda x: frozenset((x['nmvw_actor1'], x['nmvw_actor2'])) in pairs, axis=1)]

df1 = pd.DataFrame(pairs, columns=['nmvw_actor1', 'nmvw_actor2'])
len(df1)


3878

Generate [CQ3_bb_after_links.csv](CQ3_bb_after_links.csv)
```
SELECT * WHERE{
     ?nmvw_actor owl:sameAs ?bb_actor1 .
     ?bb_acq1 crm:P23_transferred_title_from ?bb_actor1 .
     ?bb_acq1 crm:P24_transferred_title_of ?bb_object .
    
    ?bb_acq2 crm:P24_transferred_title_of ?bb_object .
    ?bb_acq2 crm:P23_transferred_title_from ?bb_actor2 .

    Filter (?bb_actor1!= ?bb_actor2)
Graph <https://pressingmatter.nl/Bronbeek/Objects/Objects/assertion/bad21d24/2024-03-01T12:54>{
        ?bb_object a crm:E22_Human-Made_Object.
   }
}
```

In [191]:
df2 = pd.read_csv("CQ3_bb_after_links.csv")
df2 = df2.iloc[: , :-1]
df2.columns = [ "nmvw_actor" , "bb_actor1" , "bb_acq1" , "bb_object" , "bb_acq2" , "bb_actor2"]
pairs = set(frozenset(pair) for pair in zip(df2['bb_actor1'], df2['bb_actor2']))
df2 = pd.DataFrame(pairs, columns=['bb_actor1', 'bb_actor2'])
len(df2)

39

In [196]:
match_df = pd.read_csv("initialmatch.csv")
match_df = match_df.iloc[:, :-1]
match_df.columns = [ "nmvw_actor" , "bb_actor"]
match_df

nmvw_actor  \
0     "https://hdl.handle.net/20.500.11840/pi1021"    
1     "https://hdl.handle.net/20.500.11840/pi1060"    
2     "https://hdl.handle.net/20.500.11840/pi1064"    
3     "https://hdl.handle.net/20.500.11840/pi1099"    
4     "https://hdl.handle.net/20.500.11840/pi1111"    
..                                              ...   
973    "https://hdl.handle.net/20.500.11840/pi858"    
974    "https://hdl.handle.net/20.500.11840/pi903"    
975    "https://hdl.handle.net/20.500.11840/pi918"    
976    "https://hdl.handle.net/20.500.11840/pi918"    
977    "https://hdl.handle.net/20.500.11840/pi942"    

                                              bb_actor  
0     "https://pressingmatter.nl/Bronbeek/Constitue...  
1     "https://pressingmatter.nl/Bronbeek/Constitue...  
2     "https://pressingmatter.nl/Bronbeek/Constitue...  
3     "https://pressingmatter.nl/Bronbeek/Constitue...  
4     "https://pressingmatter.nl/Bronbeek/Constitue...  
..                                                 ...  
973   "https://pressingmatter.nl/Bronbeek/Constitue...  
974   "https://pressingmatter.nl/Bronbeek/Constitue...  
975   "https://pressingmatter.nl/Bronbeek/Constitue...  
976   "https://pressingmatter.nl/Bronbeek/Constitue...  
977   "https://pressingmatter.nl/Bronbeek/Constitue...  

[978 rows x 2 columns]

In [211]:
# Merge df1 with match_df on nmvw_actor1
merged_df1 = df1.merge(match_df, left_on='nmvw_actor1', right_on='nmvw_actor', how='inner')
# Merge df1 with match_df on nmvw_actor2
merged_df2 = df1.merge(match_df, left_on='nmvw_actor2', right_on='nmvw_actor', how='inner')
# Concatenate the merged DataFrames
final_df = pd.concat([merged_df1, merged_df2])
final_df = final_df.drop_duplicates()
# Merge final_df with df2 on bb_actor1
result_df1 = final_df.merge(df2, left_on='bb_actor', right_on='bb_actor1', how='inner')
result_df2 = final_df.merge(df2, left_on='bb_actor', right_on='bb_actor2', how='inner')
result_df = pd.concat([result_df2,result_df1])
result_df.to_csv('CQ3_result.csv', index=False)

> 10 new links from here. 

# CQ 4

##### These are the objects, where we do not know it's production place, but do know the acqusuiton actor.
> For 49152 object we did not know the production location

```PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>

SELECT (COUNT(DISTINCT ?nmvw_obj) AS ?no_object) ?nmvw_actor
WHERE{
 GRAPH <https://pressingmatter.nl/NMVW/ccrdfobj.ttl>{
    ?nmvw_acq crm:P23_transferred_title_from ?nmvw_actor.
    ?nmvw_acq crm:P24_transferred_title_of ?nmvw_obj .
    ?nmvw_obj crm:P108i_was_produced_by ?nmvw_prod .
    FILTER NOT EXISTS {
      ?nmvw_prod crm:P7_took_place_at ?nmvw_prod_place.}
    }
} GROUP BY ?nmvw_actor ```

##### These are the objects, where we DO KNOW it's production place, and acqusuiton actor.
> For 457722 object, we did not know the production location
```
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>

SELECT (COUNT(DISTINCT ?nmvw_obj) AS ?no_object) (COUNT(DISTINCT ?nmvw_prod_place) AS ?no_place) ?nmvw_actor
WHERE{
 GRAPH <https://pressingmatter.nl/NMVW/ccrdfobj.ttl>{
     ?nmvw_acq crm:P23_transferred_title_from ?nmvw_actor.
      ?nmvw_acq crm:P24_transferred_title_of ?nmvw_obj .
      ?nmvw_obj crm:P108i_was_produced_by ?nmvw_prod .
     ?nmvw_prod crm:P7_took_place_at ?nmvw_prod_place.
    }
} GROUP BY ?nmvw_actor 
```

> If we connect them with acqusistion_actor, for 45756 objects we can prioritise possible location .

##### Object connected to Bronbeek actors that as sameAs Links with NMVW
```
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>

SELECT (COUNT(DISTINCT ?bb_obj) AS ?no_object) ?bb_actor
WHERE{
    ?nmvw_actor owl:sameAs ?bb_actor .
    {?bb_acq crm:P23_transferred_title_from ?bb_actor .	
    ?bb_acq crm:P24_transferred_title_of ?bb_obj .}
    UNION{
          ?bb_obj crm:P51_has_former_or_current_owner ?bb_actor .
}

 GRAPH <https://pressingmatter.nl/Bronbeek/Objects/Objects/assertion/bad21d24/2024-03-01T12:54>{
    ?bb_obj a crm:E22_Human-Made_Object. }
} GROUP BY ?bb_actor . 

```
> total object 7996

In [ ]:
import pandas as pd

df1 = pd.read_csv("CQ4_nmvw_object_place_unknown.csv")
df2 = pd.read_csv("CQ4_nmvw_object_place_known.csv")

In [ ]:
merged_df = pd.merge(df1, df2, on='nmvw_actor', how='inner')

In [ ]:
merged_df['no_object_x'].sum()

In [ ]:
df3 = pd.read_csv("CQ4_bb_object.csv")
df3['no_object'].sum()

In [ ]:
df4 = pd.read_csv("sameAs.csv")
merged_df = pd.merge(df3, df4, on='bb_actor', how='inner')

In [ ]:
merged_df = pd.merge(merged_df, df2, on='nmvw_actor', how='inner')

In [ ]:
merged_df['no_object_x'].sum()

Now for 1841 bronbeek object we can potentially project possible location.

# CQ-5

```
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>

SELECT (AVG(?n) AS ?avg) WHERE
{
  SELECT (COUNT(DISTINCT ?nmvw_obj) AS ?n) 
  WHERE{
      GRAPH <https://pressingmatter.nl/NMVW/ccrdfhiseve.ttl>{
      ?histevet a crm:E7_Activity .
      ?histevet crm:P140i_was_attributed_by ?o .
      ?histevet  crm:P1_is_identified_by ?title .
      ?histevet crm:P4_has_time-span [crm:P82a_begin_of_the_begin ?histevet_btime;
                                       crm:P82b_end_of_the_end ?histevet_etime] .
      ?title crm:P190_has_symbolic_content ?histevet_name .
      ?o a crm:E13_Attribute_Assignment .
      ?o crm:P141_assigned ?nmvw_obj .
    } 
    ?nmvw_obj a crm:E22_Human-Made_Object .
}GROUP BY ?histevet
}
```
> Average object per event: 261.941 with 51 event

Query to download ([c5_nmvw.csv](c5_nmvw.csv))

```
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>
PREFIX owl: <http://www.w3.org/2002/07/owl#>

SELECT ?histevet ?histevet_btime ?histevet_etime ?nmvw_obj ?nmvw_actor 
  WHERE{
      GRAPH <https://pressingmatter.nl/NMVW/ccrdfhiseve.ttl>{
      ?histevet a crm:E7_Activity .
      ?histevet crm:P140i_was_attributed_by ?o .
      ?histevet  crm:P1_is_identified_by ?title .
      ?histevet crm:P4_has_time-span [crm:P82a_begin_of_the_begin ?histevet_btime;
                                       crm:P82b_end_of_the_end ?histevet_etime] .
      ?title crm:P190_has_symbolic_content ?histevet_name .
      ?o a crm:E13_Attribute_Assignment .
      ?o crm:P141_assigned ?nmvw_obj .
    } 
    ?nmvw_obj a crm:E22_Human-Made_Object .
    ?nmvw_acq crm:P24_transferred_title_of ?nmvw_obj .
    ?nmvw_acq crm:P23_transferred_title_from ?nmvw_actor .	
  	?nmvw_actor owl:sameAs ?bb_actor .
}
```

In [78]:
import pandas as pd

def cleanup_time(x: str):
    # print(type(x))
    return int(str(x).replace('"', '').strip()[:4])

df1 = pd.read_csv("c5_nmvw.csv")
df1 = df1.iloc[: , :-1]
df1.columns = [ "histevet" , "histevet_btime" , "histevet_etime" , "nmvw_obj" , "nmvw_actor"]
df1['histevet_btime'] = df1['histevet_btime'].apply(cleanup_time)
df1['histevet_etime'] = df1['histevet_etime'].apply(cleanup_time)

# drop the nmvw_obj
df1 = df1.drop(['nmvw_obj'], axis=1)
# Remove duplicate rows in df2
df1 = df1.drop_duplicates()
df1

histevet  histevet_btime  \
0       "https://hdl.handle.net/20.500.11840/event31"             1935   
64      "https://hdl.handle.net/20.500.11840/event11"             1906   
96      "https://hdl.handle.net/20.500.11840/event13"             1920   
240     "https://hdl.handle.net/20.500.11840/event15"             1936   
330    "https://hdl.handle.net/20.500.11840/event423"             1873   
331    "https://hdl.handle.net/20.500.11840/event423"             1873   
333    "https://hdl.handle.net/20.500.11840/event423"             1873   
339    "https://hdl.handle.net/20.500.11840/event423"             1873   
383    "https://hdl.handle.net/20.500.11840/event423"             1873   
416    "https://hdl.handle.net/20.500.11840/event423"             1873   
427    "https://hdl.handle.net/20.500.11840/event423"             1873   
442    "https://hdl.handle.net/20.500.11840/event423"             1873   
487    "https://hdl.handle.net/20.500.11840/event423"             1873   
498    "https://hdl.handle.net/20.500.11840/event423"             1873   
528    "https://hdl.handle.net/20.500.11840/event423"             1873   
663    "https://hdl.handle.net/20.500.11840/event423"             1873   
670    "https://hdl.handle.net/20.500.11840/event423"             1873   
698    "https://hdl.handle.net/20.500.11840/event423"             1873   
702    "https://hdl.handle.net/20.500.11840/event423"             1873   
725    "https://hdl.handle.net/20.500.11840/event423"             1873   
772    "https://hdl.handle.net/20.500.11840/event423"             1873   
817    "https://hdl.handle.net/20.500.11840/event423"             1873   
831    "https://hdl.handle.net/20.500.11840/event423"             1873   
844    "https://hdl.handle.net/20.500.11840/event423"             1873   
946    "https://hdl.handle.net/20.500.11840/event423"             1873   
1064   "https://hdl.handle.net/20.500.11840/event423"             1873   
1269   "https://hdl.handle.net/20.500.11840/event423"             1873   
1296   "https://hdl.handle.net/20.500.11840/event423"             1873   
1343   "https://hdl.handle.net/20.500.11840/event424"             1859   
1518   "https://hdl.handle.net/20.500.11840/event424"             1859   
1561   "https://hdl.handle.net/20.500.11840/event425"             1872   
1563   "https://hdl.handle.net/20.500.11840/event425"             1872   
1585   "https://hdl.handle.net/20.500.11840/event426"             1900   
1601   "https://hdl.handle.net/20.500.11840/event427"             1904   
1602   "https://hdl.handle.net/20.500.11840/event427"             1904   
1604   "https://hdl.handle.net/20.500.11840/event427"             1904   
1607   "https://hdl.handle.net/20.500.11840/event430"             1945   
1632   "https://hdl.handle.net/20.500.11840/event433"             1894   
1633    "https://hdl.handle.net/20.500.11840/event30"             1925   

      histevet_etime                                       nmvw_actor  
0               1936     "https://hdl.handle.net/20.500.11840/pi703"   
64              1906    "https://hdl.handle.net/20.500.11840/pi2664"   
96              1921     "https://hdl.handle.net/20.500.11840/pi703"   
240             1936    "https://hdl.handle.net/20.500.11840/pi6890"   
330             1942   "https://hdl.handle.net/20.500.11840/pi21230"   
331             1942    "https://hdl.handle.net/20.500.11840/pi3325"   
333             1942    "https://hdl.handle.net/20.500.11840/pi7750"   
339             1942   "https://hdl.handle.net/20.500.11840/pi56724"   
383             1942    "https://hdl.handle.net/20.500.11840/pi3581"   
416             1942    "https://hdl.handle.net/20.500.11840/pi3153"   
427             1942   "https://hdl.handle.net/20.500.11840/pi81966"   
442             1942    "https://hdl.handle.net/20.500.11840/pi4862"   
487             1942    "https://hdl.handle.net/20.500.11840/pi5957"   
498             1942     "https://hdl.handle.net/20.500.11840/pi499"   
528  

In [83]:
def cleanup_time(x: str):
    # Check if x is not None and not an empty string
    try:
        if x == "None" :
            return 0
        if x and str(x).strip():
            # Convert the cleaned string to an integer
            return int(str(x).replace('"', '').strip()[:4])
    except ValueError:
        return 0

# Read the second CSV file into DataFrame df2
df2 = pd.read_csv("c5_bb.csv")
df2 = df2.iloc[: , :-1]
df2.columns = ["bb_obj" , "bb_actor" , "nmvw_actor" , "b_time" , "e_time"]
df2['b_time'] = df2['b_time'].apply(cleanup_time)
df2['e_time'] = df2['e_time'].apply(cleanup_time)

# Remove duplicate rows in df2
df2 = df2.drop_duplicates() 

# Extract unique 'nmvw_actor' values from df2
unique_nmvw_actor = df1['nmvw_actor'].unique()

# Filter df1 to keep only rows where 'nmvw_actor' exists in df2
df1_filtered = df1[df1['nmvw_actor'].isin(unique_nmvw_actor)]

df2 = df2.sort_values('nmvw_actor')

df2

bb_obj  \
0      "https://pressingmatter.nl/Bronbeek/Objects/5...   
1      "https://pressingmatter.nl/Bronbeek/Objects/5...   
2      "https://pressingmatter.nl/Bronbeek/Objects/5...   
3      "https://pressingmatter.nl/Bronbeek/Objects/5...   
4      "https://pressingmatter.nl/Bronbeek/Objects/5...   
...                                                 ...   
3280   "https://pressingmatter.nl/Bronbeek/Objects/1...   
3281   "https://pressingmatter.nl/Bronbeek/Objects/1...   
3282   "https://pressingmatter.nl/Bronbeek/Objects/1...   
3283   "https://pressingmatter.nl/Bronbeek/Objects/3...   
3284   "https://pressingmatter.nl/Bronbeek/Objects/5...   

                                               bb_actor  \
0      "https://pressingmatter.nl/Bronbeek/Constitue...   
1      "https://pressingmatter.nl/Bronbeek/Constitue...   
2      "https://pressingmatter.nl/Bronbeek/Constitue...   
3      "https://pressingmatter.nl/Bronbeek/Constitue...   
4      "https://pressingmatter.nl/Bronbeek/Constitue...   
...                                                 ...   
3280   "https://pressingmatter.nl/Bronbeek/Constitue...   
3281   "https://pressingmatter.nl/Bronbeek/Constitue...   
3282   "https://pressingmatter.nl/Bronbeek/Constitue...   
3283   "https://pressingmatter.nl/Bronbeek/Constitue...   
3284   "https://pressingmatter.nl/Bronbeek/Constitue...   

                                          nmvw_actor  b_time  e_time  
0      "https://hdl.handle.net/20.500.11840/pi1064"     1898    1898  
1      "https://hdl.handle.net/20.500.11840/pi1064"     1898    1898  
2      "https://hdl.handle.net/20.500.11840/pi1171"     1922    1922  
3      "https://hdl.handle.net/20.500.11840/pi1171"     1926    1926  
4      "https://hdl.handle.net/20.500.11840/pi1171"     1926    1926  
...                                              ...     ...     ...  
3280    "https://hdl.handle.net/20.500.11840/pi903"     1993    1993  
3281    "https://hdl.handle.net/20.500.11840/pi903"     1993    1993  
3282    "https://hdl.handle.net/20.500.11840/pi903"     1993    1993  
3283    "https://hdl.handle.net/20.500.11840/pi918"     2005    2005  
3284    "https://hdl.handle.net/20.500.11840/pi942"     2011    2011  

[3285 rows x 5 columns]

In [88]:
df2['bb_obj'].nunique()

print(f"{df2['bb_obj'].nunique()} object has connection with actors where actor has connection with historical event")

2904 object has connection with actors where actor has connection with historical event


# CQ-6

```
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX aat: <http://vocab.getty.edu/aat/>

SELECT ?nmvw_obj ?nmvw_prod_place ?nmvw_acq ?b_time ?e_time
WHERE{
 GRAPH <https://pressingmatter.nl/NMVW/ccrdfobj.ttl>{
     ?nmvw_acq crm:P23_transferred_title_from ?nmvw_actor.
     ?nmvw_acq crm:P24_transferred_title_of ?nmvw_obj .
     ?nmvw_obj crm:P108i_was_produced_by ?nmvw_prod .
     ?nmvw_prod crm:P7_took_place_at ?nmvw_prod_place.
     ?prov_activity ?p ?nmvw_acq.
     ?prov_activity crm:P2_has_type aat:300055863 .
     ?prov_activity crm:P4_has_time-span ?time .
     ?time crm:P82a_begin_of_the_begin ?b_time .
     ?time crm:P82b_end_of_the_end ?e_time.
    }
}
```

In [44]:
def cleanup_time(x: str):
    return int(str(x).replace('"', '').strip()[-4:])

import pandas as pd

df1 = pd.read_csv("c6_nmvw.csv")
df1 = df1.iloc[: , :-1]
df1.columns = ['nmvw_obj', 'nmvw_prod_place', 'nmvw_actor', 'b_time', 'e_time']
df1['b_time'] = df1['b_time'].apply(cleanup_time)
df1['e_time'] = df1['e_time'].apply(cleanup_time)
df1



nmvw_obj  \
0      "https://hdl.handle.net/20.500.11840/128787"    
1      "https://hdl.handle.net/20.500.11840/283554"    
2      "https://hdl.handle.net/20.500.11840/143099"    
3      "https://hdl.handle.net/20.500.11840/302925"    
4      "https://hdl.handle.net/20.500.11840/302926"    
...                                              ...   
3531   "https://hdl.handle.net/20.500.11840/235556"    
3532   "https://hdl.handle.net/20.500.11840/235556"    
3533   "https://hdl.handle.net/20.500.11840/235557"    
3534   "https://hdl.handle.net/20.500.11840/235557"    
3535   "https://hdl.handle.net/20.500.11840/235558"    

                                        nmvw_prod_place  \
0      "https://hdl.handle.net/20.500.11840/termmast...   
1      "https://hdl.handle.net/20.500.11840/termmast...   
2      "https://hdl.handle.net/20.500.11840/termmast...   
3      "https://hdl.handle.net/20.500.11840/termmast...   
4      "https://hdl.handle.net/20.500.11840/termmast...   
...                                                 ...   
3531   "https://hdl.handle.net/20.500.11840/termmast...   
3532   "https://hdl.handle.net/20.500.11840/termmast...   
3533   "https://hdl.handle.net/20.500.11840/termmast...   
3534   "https://hdl.handle.net/20.500.11840/termmast...   
3535   "https://hdl.handle.net/20.500.11840/termmast...   

                                           nmvw_actor  b_time  e_time  
0       "https://hdl.handle.net/20.500.11840/pi2338"     1972    1972  
1       "https://hdl.handle.net/20.500.11840/pi3989"     1969    1969  
2       "https://hdl.handle.net/20.500.11840/pi3683"     1978    1978  
3      "https://hdl.handle.net/20.500.11840/pi38456"     2003    2008  
4      "https://hdl.handle.net/20.500.11840/pi38456"     2003    2008  
...                                               ...     ...     ...  
3531   "https://hdl.handle.net/20.500.11840/pi38479"     2003    2003  
3532   "https://hdl.handle.net/20.500.11840/pi38479"     2003    2003  
3533   "https://hdl.handle.net/20.500.11840/pi38479"     2003    2003  
3534   "https://hdl.handle.net/20.500.11840/pi38479"     2003    2003  
3535   "https://hdl.handle.net/20.500.11840/pi38479"     2003    2003  

[3536 rows x 5 columns]

In [59]:
grouped_df = df1.groupby(['nmvw_actor', 'b_time', 'e_time'])['nmvw_obj'].nunique().reset_index()
len(grouped_df['nmvw_obj'])

137

```
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX crm: <http://www.cidoc-crm.org/cidoc-crm/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX aat: <http://vocab.getty.edu/aat/>

SELECT ?bb_obj ?bb_actor ?nmvw_actor ?b_time ?e_time
WHERE{
  	?time crm:P82a_begin_of_the_begin ?b_time .
    ?time crm:P82b_end_of_the_end ?e_time.
  	?prov_activity crm:P9_consists_of ?bb_acq .
  	?prov_activity crm:P2_has_type aat:300055863 .
  	?prov_activity  crm:P9_consists_of ?bb_acq2 .
  	?bb_acq2 crm:P23_transferred_title_from ?bb_actor . 
  	?nmvw_actor owl:sameAs ?bb_actor .
    ?bb_acq crm:P24_transferred_title_of ?bb_obj .
  FILTER (?bb_acq != ?bb_acq2) .

 GRAPH <https://pressingmatter.nl/Bronbeek/Objects/Objects/assertion/bad21d24/2024-03-01T12:54>{
    ?bb_obj a crm:E22_Human-Made_Object. }
}
```

In [42]:
def cleanup_time(x: str):
    # Check if x is not None and not an empty string
    try:
        if x == "None" :
            return 0
        if x and str(x).strip():
            # Convert the cleaned string to an integer
            return int(str(x).replace('"', '').strip()[:4])
    except ValueError:
        return 0

import pandas as pd

df2 = pd.read_csv("c6_bb.csv")
df2 = df2.iloc[: , :-1]
df2.columns = ['bb_obj', 'bb_actor', 'nmvw_actor', 'b_time', 'e_time']
df2['b_time'] = df2['b_time'].apply(cleanup_time)
df2['e_time'] = df2['e_time'].apply(cleanup_time)
df2

bb_obj  \
0      "https://pressingmatter.nl/Bronbeek/Objects/5...   
1      "https://pressingmatter.nl/Bronbeek/Objects/5...   
2      "https://pressingmatter.nl/Bronbeek/Objects/5...   
3      "https://pressingmatter.nl/Bronbeek/Objects/5...   
4      "https://pressingmatter.nl/Bronbeek/Objects/5...   
...                                                 ...   
3280   "https://pressingmatter.nl/Bronbeek/Objects/1...   
3281   "https://pressingmatter.nl/Bronbeek/Objects/1...   
3282   "https://pressingmatter.nl/Bronbeek/Objects/1...   
3283   "https://pressingmatter.nl/Bronbeek/Objects/3...   
3284   "https://pressingmatter.nl/Bronbeek/Objects/5...   

                                               bb_actor  \
0      "https://pressingmatter.nl/Bronbeek/Constitue...   
1      "https://pressingmatter.nl/Bronbeek/Constitue...   
2      "https://pressingmatter.nl/Bronbeek/Constitue...   
3      "https://pressingmatter.nl/Bronbeek/Constitue...   
4      "https://pressingmatter.nl/Bronbeek/Constitue...   
...                                                 ...   
3280   "https://pressingmatter.nl/Bronbeek/Constitue...   
3281   "https://pressingmatter.nl/Bronbeek/Constitue...   
3282   "https://pressingmatter.nl/Bronbeek/Constitue...   
3283   "https://pressingmatter.nl/Bronbeek/Constitue...   
3284   "https://pressingmatter.nl/Bronbeek/Constitue...   

                                          nmvw_actor  b_time  e_time  
0      "https://hdl.handle.net/20.500.11840/pi1064"     1898    1898  
1      "https://hdl.handle.net/20.500.11840/pi1064"     1898    1898  
2      "https://hdl.handle.net/20.500.11840/pi1171"     1922    1922  
3      "https://hdl.handle.net/20.500.11840/pi1171"     1926    1926  
4      "https://hdl.handle.net/20.500.11840/pi1171"     1926    1926  
...                                              ...     ...     ...  
3280    "https://hdl.handle.net/20.500.11840/pi903"     1993    1993  
3281    "https://hdl.handle.net/20.500.11840/pi903"     1993    1993  
3282    "https://hdl.handle.net/20.500.11840/pi903"     1993    1993  
3283    "https://hdl.handle.net/20.500.11840/pi918"     2005    2005  
3284    "https://hdl.handle.net/20.500.11840/pi942"     2011    2011  

[3285 rows x 5 columns]

In [57]:
def merge_with_condition(df1, df2):
    merged_df = pd.merge(df1, df2, on='nmvw_actor', how='inner')
    merged_df['time_diff_b'] = merged_df['b_time_y'] - merged_df['b_time_x']
    merged_df['time_diff_e'] = merged_df['e_time_y'] - merged_df['e_time_x']
    condition = (abs(merged_df['time_diff_b']) <= 20) & (abs(merged_df['time_diff_e']) <= 20)
    return merged_df[condition]

# Merge the dataframes based on the custom condition
result_df = merge_with_condition(df1, df2)
print(result_df.to_csv(index=False))
print("Number of rows in result_df:", len(result_df))

nmvw_obj,nmvw_prod_place,nmvw_actor,b_time_x,e_time_x,bb_obj,bb_actor,b_time_y,e_time_y,time_diff_b,time_diff_e
" ""https://hdl.handle.net/20.500.11840/209379"" "," ""https://hdl.handle.net/20.500.11840/termmaster7055"" "," ""https://hdl.handle.net/20.500.11840/pi1064"" ",1915,1915," ""https://pressingmatter.nl/Bronbeek/Objects/5148"" "," ""https://pressingmatter.nl/Bronbeek/Constituents/191"" ",1898,1898,-17,-17
" ""https://hdl.handle.net/20.500.11840/209379"" "," ""https://hdl.handle.net/20.500.11840/termmaster7055"" "," ""https://hdl.handle.net/20.500.11840/pi1064"" ",1915,1915," ""https://pressingmatter.nl/Bronbeek/Objects/5149"" "," ""https://pressingmatter.nl/Bronbeek/Constituents/191"" ",1898,1898,-17,-17
" ""https://hdl.handle.net/20.500.11840/209382"" "," ""https://hdl.handle.net/20.500.11840/termmaster7057"" "," ""https://hdl.handle.net/20.500.11840/pi1064"" ",1915,1915," ""https://pressingmatter.nl/Bronbeek/Objects/5148"" "," ""https://pressingmatter.nl/Bronbeek/Constituents